lib/common/socket.c

```diff
@@ -1652,6 +1887,18 @@ static void proceed_handshake(h2o_socket_t *sock, const char *err)
         return;
     }
 
+    if (init_delayed_sockets()) {
+        if (async_handshake_in_flight != NULL && async_handshake_in_flight != sock) {
+            // handshake is in-flight, delay the handshake
+            if (!h2o_linklist_is_linked(&sock->async.delayed_link)) {
+                sock->async.delayed_cb = retry_proceed_handshake;
+                // fprintf(stderr, "retry handshake\n");
+                h2o_linklist_insert(&delayed_sockets, &sock->async.delayed_link);
+            }
+            return;
+        }
+    }
```

Comment:
Am I correct to understand that the intent of having the "delayed sockets" structure is to limit the RSA operation inflight to one at a time? Because otherwise, write to the neverbleed socket might block?

Assuming that is the case, I'm not sure if this is the adequate way of avoiding such blocking.

The problem with this approach is that it blocks all types of TLS hanhshakes (e.g., resumptions, full handshakes using ECSDA certificates) from making progress when an RSA signature generation is inflight. I do not think we'd want to those very lightweight handshakes to be blocked while a heavyweight operation like RSA signature generation is inflight.

As a way forward, I wonder if we could do request pipelining on the neverbleed socket (i.e., write as many requests as possible, then read one by one as results become available).

Reply:
Fully agreed, this is meant as an interim solution.

I have a branch in-progress/testing for pipelining.